gax-java/gax/src/main/java/com/google/api/gax/rpc/RequestParamsBuilder.java

```diff
@@ -56,19 +59,49 @@ public static RequestParamsBuilder create() {
    * pre-configured path templates. This method is called repeatedly for each configured routing
    * rule parameter, it's possible that the incoming field value from request is null or there is no
    * matches found, we'll continue the match-and-extract process for the next routing rule parameter
-   * in such case.
+   * in such case. This method will percent encode both the header key and header value.
    *
    * @param fieldValue the field value from a request
    * @param headerKey the header key for the routing header param
    * @param pathTemplate {@link PathTemplate} the path template used for match-and-extract
    */
   public void add(String fieldValue, String headerKey, PathTemplate pathTemplate) {
-    if (fieldValue == null) {
+    if (fieldValue == null || fieldValue.isEmpty()) {
```

Comment:
Adding this b/c it failed the showcase test. Checks that the value is non-null and the matched value is non-empty.

Reply:
matchedValues should be empty if `fieldValue.isEmpty()`, which line is failing in showcase test without this change?

Reply:
There are a few tests that fail (both transports). For example: `testGrpc_matchesOtherHeaderName`

This sets the Other_Header to be `instances/456` and would expect that the header to only be `baz=instances%2F456`.

The result of the header is: `header=&routing_id=&baz=instances%2F456` as it also matches these two routing header values: https://github.com/googleapis/gapic-showcase/blob/c41cdbd3a8ebf3c30ffa00dbeee1d96a6ae5fbcc/schema/google/showcase/v1beta1/echo.proto#L52-L58

From the routing.proto, I don't think sending an empty string is beneficial: https://github.com/googleapis/googleapis/blob/53a0be29c4a95a1d3b4c0d3a7a2ac8b52af2a3c0/google/api/routing.proto#L109-L110

Reply:
I'm kind of surprised the result is `header=&routing_id=&baz=instances%2F456`. I agree with the expected behavior, but I would expect the current behavior should match the expected behavior because of this logic 
https://github.com/googleapis/sdk-platform-java/blob/e7b2dbf24c4ecad5948340bc6078818167c059d5/gax-java/gax/src/main/java/com/google/api/gax/rpc/RequestParamsBuilder.java#L73
`matchedValues` would be empty if `fieldValue` is empty and I just verified with unit tests, so theoretically no headers should be added to the builder.

Reply:
It seems to be with some specific cases (i.e. not setting a header value when the pathtemplate is {header=**}). The matchedValues map has one entry (header -> "") and it'll match anything. I believe for most cases the logic will work properly.

Reply:
I see, that makes sense and good to know. So a test case like this would fail without this change? 
https://github.com/googleapis/sdk-platform-java/blob/7b6c9522fa0c16c5f98507bae9afff9a05bf7882/gax-java/gax/src/test/java/com/google/api/gax/rpc/RequestParamsBuilderTest.java#L136-L140

Reply:
Took another look at this case and I'm not sure it's the best test case for scenario above. I'll rename to test case to reflect what I think is causing this: protobuf's default value for strings. I think the test case should be: 

```
    Map<String, String> actual = getRoutingHeaders("{table_location=**}", "");
    assertThat(actual).isEmpty();
```
where protobuf puts `""` as the value when table_location routing param wasn't set in the request message.